cocotb/decorators.py

```diff
@@ -446,7 +450,21 @@ def __init__(self, f, timeout=None, expect_fail=False, expect_error=False,
 
     def __call__(self, *args, **kwargs):
         try:
-            return RunningTest(self._func(*args, **kwargs), self)
+            rc = coroutine(self._func)(*args, **kwargs)
+
+            @wraps(self._func)
+            def wrapper():
+                tst = cocotb.fork(rc)
+
+                if self.timeout is None:
+                    yield tst.join()
+                else:
+                    tmr = Timer(self.timeout, units=self.timeout_units)
+                    res = yield [tst.join(), tmr]
+                    if res == tmr:
+                        raise TestFailure("The test runtime exceeded the timeout of %d%s}" % (self.timeout, self.timeout_units))
+
+            self._func = wrapper
```

Comment:
This wrapping should happen in `__init__`, not here. In the unlikely event that `__call__` runs twice, this is going to behave poorly

Reply:
Originally that is where I had it, but that causes other problems.  My goal was not to mask any issues, which is why I create the running coroutine outside of the wrapper function, so it would catch the errors of not being a coroutine and be treated as it always has been.  If I tried to handle that in the `__init__`, the code would either mask the error, changing from the test not being run and listed as n/a in the results, to a test with a `TestError` and not `TestFailure`.

What if `__call__` were implemented like:
```python

    def __call__(self, *args, **kwargs):
        try:
            rc = coroutine(self._func)(*args, **kwargs)

            @wraps(self._func)
            def wrapper():
                tst = cocotb.fork(rc)

                if self.timeout is None:
                    yield tst.join()
                else:
                    tmr = Timer(self.timeout, units=self.timeout_units)
                    res = yield [tst.join(), tmr]
                    if res == tmr:
                        raise TestFailure("The test runtime exceeded the timeout of %d%s}" % (self.timeout, self.timeout_units))

            return RunningTest(wrapper(), self)
        except Exception as e:
            raise raise_error(self, "Test raised exception:")
```

By removing the line to save the wrapper function, that will prevent things from behaving poorly if `__call__` were to be called twice.